### PR TITLE
fix testcases while chrono-tz enabled

### LINKED ISSUE #2931: failed to pass test cases while enabling feature chrono-tz
```bash
cargo test --features chrono-tz
```

```bash
failures:

---- compute::kernels::cast::tests::test_cast_timestamp_to_string stdout ----
[arrow/src/compute/kernels/cast.rs:3852] &array = PrimitiveArray<Timestamp(Millisecond, None)>
[
  1997-05-19T00:00:00.005,
  2018-12-25T00:00:00.001,
  null,
]
thread 'compute::kernels::cast::tests::test_cast_timestamp_to_string' panicked at 'assertion failed: `(left == right)`
  left: `"1997-05-19 00:00:00.005"`,
 right: `"1997-05-19 00:00:00.005 +00:00"`', arrow/src/compute/kernels/cast.rs:3856:9

---- compute::kernels::cast::tests::test_timestamp_cast_utf8 stdout ----
thread 'compute::kernels::cast::tests::test_timestamp_cast_utf8' panicked at 'assertion failed: `(left == right)`
  left: `StringArray
[
  "1970-01-01 20:30:00 +10:00",
  null,
  "1970-01-02 09:58:59 +10:00",
]`,
 right: `StringArray
[
  "1970-01-01 20:30:00",
  null,
  "1970-01-02 09:58:59",
]`', arrow/src/compute/kernels/cast.rs:5762:9

---- csv::writer::tests::test_export_csv_timestamps stdout ----
thread 'csv::writer::tests::test_export_csv_timestamps' panicked at 'assertion failed: `(left == right)`
  left: `Some("c1,c2\n2019-04-18T20:54:47.378000000+10:00,2019-04-18T10:54:47.378000000+00:00\n2021-10-30T17:59:07.000000000+11:00,2021-10-30T06:59:07.000000000+00:00\n")`,
 right: `Some("c1,c2\n2019-04-18T20:54:47.378000000+10:00,2019-04-18T10:54:47.378000000\n2021-10-30T17:59:07.000000000+11:00,2021-10-30T06:59:07.000000000\n")`', arrow/src/csv/writer.rs:652:9

---- csv::writer::tests::test_write_csv stdout ----
thread 'csv::writer::tests::test_write_csv' panicked at 'assertion failed: `(left == right)`
  left: `"c1,c2,c3,c4,c5,c6,c7\nLorem ipsum dolor sit amet,123.564532,3,true,,00:20:34,cupcakes\nconsectetur adipiscing elit,,2,false,2019-04-18T10:54:47.378000000+00:00,06:51:20,cupcakes\nsed do eiusmod tempor,-556132.25,1,,2019-04-18T02:45:55.555000000+00:00,23:46:03,foo\nLorem ipsum dolor sit amet,123.564532,3,true,,00:20:34,cupcakes\nconsectetur adipiscing elit,,2,false,2019-04-18T10:54:47.378000000+00:00,06:51:20,cupcakes\nsed do eiusmod tempor,-556132.25,1,,2019-04-18T02:45:55.555000000+00:00,23:46:03,foo\n"`,
 right: `"c1,c2,c3,c4,c5,c6,c7\nLorem ipsum dolor sit amet,123.564532,3,true,,00:20:34,cupcakes\nconsectetur adipiscing elit,,2,false,2019-04-18T10:54:47.378000000,06:51:20,cupcakes\nsed do eiusmod tempor,-556132.25,1,,2019-04-18T02:45:55.555000000,23:46:03,foo\nLorem ipsum dolor sit amet,123.564532,3,true,,00:20:34,cupcakes\nconsectetur adipiscing elit,,2,false,2019-04-18T10:54:47.378000000,06:51:20,cupcakes\nsed do eiusmod tempor,-556132.25,1,,2019-04-18T02:45:55.555000000,23:46:03,foo\n"`', arrow/src/csv/writer.rs:546:9


failures:
    compute::kernels::cast::tests::test_cast_timestamp_to_string
    compute::kernels::cast::tests::test_timestamp_cast_utf8
    csv::writer::tests::test_export_csv_timestamps
    csv::writer::tests::test_write_csv

test result: FAILED. 751 passed; 4 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.21s
```

### PR DESCRIPTION
**Pull request by waitingkuo**
# Which issue does this PR close?

Closes #2931 

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/compute/kernels/cast.rs [hunk 1]
```diff
@@ -3853,8 +3853,8 @@ mod tests {
         let b = cast(&array, &DataType::Utf8).unwrap();
         let c = b.as_any().downcast_ref::<StringArray>().unwrap();
         assert_eq!(&DataType::Utf8, c.data_type());
-        assert_eq!("1997-05-19 00:00:00.005", c.value(0));
-        assert_eq!("2018-12-25 00:00:00.001", c.value(1));
+        assert_eq!("1997-05-19 00:00:00.005 +00:00", c.value(0));
+        assert_eq!("2018-12-25 00:00:00.001 +00:00", c.value(1));
```
Comment (waitingkuo) [inline arrow/src/compute/kernels/cast.rs:3857]: the result for `Timestamp(Unit, Some("UTC"))` should contains +00:00

File: arrow/src/compute/kernels/cast.rs [hunk 2]
```diff
@@ -5754,9 +5754,9 @@ mod tests {
         let out = cast(&(Arc::new(array) as ArrayRef), &DataType::Utf8).unwrap();
 
         let expected = StringArray::from(vec![
-            Some("1970-01-01 20:30:00"),
+            Some("1970-01-01 20:30:00 +10:00"),
             None,
-            Some("1970-01-02 09:58:59"),
+            Some("1970-01-02 09:58:59 +10:00"),
```
Comment (waitingkuo) [inline arrow/src/compute/kernels/cast.rs:5759]: the result for Timestamp(Unit, Some("Australia/Sydnet")) should contains +10:00

File: arrow/src/csv/writer.rs [hunk 3]
```diff
@@ -527,11 +527,11 @@ mod tests {
...
 Lorem ipsum dolor sit amet,123.564532,3,true,,00:20:34,cupcakes
-consectetur adipiscing elit,,2,false,2019-04-18T10:54:47.378000000+00:00,06:51:20,cupcakes
-sed do eiusmod tempor,-556132.25,1,,2019-04-18T02:45:55.555000000+00:00,23:46:03,foo
+consectetur adipiscing elit,,2,false,2019-04-18T10:54:47.378000000,06:51:20,cupcakes
+sed do eiusmod tempor,-556132.25,1,,2019-04-18T02:45:55.555000000,23:46:03,foo
 Lorem ipsum dolor sit amet,123.564532,3,true,,00:20:34,cupcakes
-consectetur adipiscing elit,,2,false,2019-04-18T10:54:47.378000000+00:00,06:51:20,cupcakes
-sed do eiusmod tempor,-556132.25,1,,2019-04-18T02:45:55.555000000+00:00,23:46:03,foo
+consectetur adipiscing elit,,2,false,2019-04-18T10:54:47.378000000,06:51:20,cupcakes
+sed do eiusmod tempor,-556132.25,1,,2019-04-18T02:45:55.555000000,23:46:03,foo
```
Comment (waitingkuo) [inline arrow/src/csv/writer.rs:534]: `Timestamp<Unit, None>` contains no timezone

File: arrow/src/csv/writer.rs [hunk 4]
```diff
@@ -646,8 +646,8 @@ sed do eiusmod tempor,-556132.25,1,,2019-04-18T02:45:55.555000000,23:46:03,foo
         }
 
         let left = "c1,c2
-2019-04-18T20:54:47.378000000+10:00,2019-04-18T10:54:47.378000000+00:00
-2021-10-30T17:59:07.000000000+11:00,2021-10-30T06:59:07.000000000+00:00\n";
+2019-04-18T20:54:47.378000000+10:00,2019-04-18T10:54:47.378000000
+2021-10-30T17:59:07.000000000+11:00,2021-10-30T06:59:07.000000000\n";
```
Comment (waitingkuo) [inline arrow/src/csv/writer.rs:650]: Timestamp<Unit, None> contains no timezone

**Comment by tustvold** [comment]
Could we possibly add checks for this to CI?

File: .github/workflows/arrow.yml [hunk 5]
```diff
@@ -68,6 +68,8 @@ jobs:
         run: cargo test -p arrow-integration-test --all-features
       - name: Test arrow
         run: cargo test -p arrow
+      - name: Test arrow with all features
+        run: cargo test -p arrow --all-features
```
Comment (waitingkuo) [inline .github/workflows/arrow.yml:72]: @tustvold i added cargo test with all features

Comment (tustvold) [reply .github/workflows/arrow.yml:72]: I think this will fail on stable, as it will enable the SIMD feature

**Review by tustvold — APPROVED** [review]
LGTM - thank you

File: .github/workflows/arrow.yml [hunk 6]
```diff
@@ -68,9 +68,9 @@ jobs:
         run: cargo test -p arrow-integration-test --all-features
       - name: Test arrow
         run: cargo test -p arrow
-      - name: Test --features=force_validate,prettyprint,ipc_compression,ffi,dyn_cmp_dict,dyn_arith_dict
+      - name: Test --features=force_validate,prettyprint,ipc_compression,ffi,dyn_cmp_dict,dyn_arith_dict,chrono-tz
         run: |
-          cargo test -p arrow --features=force_validate,prettyprint,ipc_compression,ffi,dyn_cmp_dict,dyn_arith_dict
+          cargo test -p arrow --features=force_validate,prettyprint,ipc_compression,ffi,dyn_cmp_dict,dyn_arith_dict,chrono-tz
```
Comment (waitingkuo) [inline .github/workflows/arrow.yml:73]: @tustvold You're right, it failed while all features enabled. I added chrono-tz here instead

File: arrow/src/compute/kernels/cast.rs (hunk 1 above)
Comment (jhorstmann) [reply arrow/src/compute/kernels/cast.rs:3857]: Is the space before the timezone offset expected here?

Comment (tustvold) [reply arrow/src/compute/kernels/cast.rs:3857]: It appears to output https://docs.rs/chrono/latest/chrono/struct.DateTime.html#method.to_rfc2822 which has spaces

Comment (tustvold) [reply arrow/src/compute/kernels/cast.rs:3857]: Filed - https://github.com/apache/arrow-rs/issues/2934

**Comment by tustvold** [comment]
Clippy seems to be finding more code to complain about now :cry:

File: arrow/src/compute/kernels/cast.rs (hunk 1 above)
Comment (waitingkuo) [reply arrow/src/compute/kernels/cast.rs:3857]: i'm working on #2917 and found this issue as well.  Is `2018-01-26T18:30:09.453829+08:00`  preferred?

Comment (jhorstmann) [reply arrow/src/compute/kernels/cast.rs:3857]: That looks like [rfc3339 format](https://www.rfc-editor.org/rfc/rfc3339#section-5.6), which is probably the most common or standardized format.

[`chrono` has a separate function for this format](https://docs.rs/chrono/latest/chrono/struct.DateTime.html#method.to_rfc3339_opts) which also allows specifying how many fractional digits you want.

**Comment by waitingkuo** [comment]
> Clippy seems to be finding more code to complain about now 😢

I just remove chrono-tz from clippy. I can submitted another pr to fix that and add it in clippy later

**Pull request merged**